### PR TITLE
pump: Fix #560, retry connection and request when notifying drainers

### LINKED ISSUE #560: pump should retry notifying registered drainers instead of exiting on startup
When pump starts, it seems to try only one time to notify registered drainers, and it exits if it fails:

```
2019/04/24 12:14:30 version.go:33: [info] Release Version: v3.0.0-beta.1-31-g1a25971
2019/04/24 12:14:30 version.go:34: [info] Git Commit Hash: 1a259716e0aaef328f94e124e02e90be58d756dc
2019/04/24 12:14:30 version.go:35: [info] Build TS: 2019-04-23 12:42:17
2019/04/24 12:14:30 version.go:36: [info] Go Version: go1.11.2
2019/04/24 12:14:30 version.go:37: [info] Go OS/Arch: linuxamd64
2019/04/24 12:14:32 server.go:116: [info] clusterID of pump server is 6683140680737258511
2019/04/24 12:14:32 storage.go:116: [info] options: &{ValueLogFileSize:524288000 Sync:true KVConfig:<nil>}
2019/04/24 12:14:32 storage.go:1090: [info] Storage config: &{BlockCacheCapacity:8388608 BlockRestartInterval:16 BlockSize:4096 CompactionL0Trigger:8 CompactionTableSize:67108864 CompactionTotalSize:536870912 CompactionTotalSizeMultiplier:8 WriteBuffer:67108864 WriteL0PauseTrigger:24 WriteL0SlowdownTrigger:17}
2019/04/24 12:14:32 storage.go:188: [info] gcTS: 407757428283932672, maxCommitTS: 407928127082725377, headPointer: {Fid:1 Offset:2000}, handlePointer: {Fid:1 Offset:2000}
2019/04/24 12:14:32 server.go:316: [info] register success, this pump's node id is localhost.localdomain:8250
2019/04/24 12:14:32 node.go:160: [info] start try to notify drainer:  192.168.236.157:8249
2019/04/24 12:14:32 main.go:71: [error] pump server error, fail to notify all living drainer: notify drainer(192.168.236.157:8249); but return error(rpc error: code = Unavailable desc = all SubConns are in TransientFailure, latest connection error: connection error: desc = "transport: Error while dialing dial tcp 192.168.236.157:8249: connect: connection refused")
```

pump should retry for some time period of number of retries before exiting. This would greatly simplify cluster startup, and would follow the example of other components that handle startup order more flexibly.

### PR DESCRIPTION
**Pull request by suzaku**
### What problem does this PR solve? 

Fix #560

### What is changed and how it works?

Retry connection and request.

### Check List 

Tests 

 - Unit test

Code changes

Side effects

Related changes

### CODE REVIEW

File: pkg/util/util_test.go [hunk 1]
```diff
@@ -276,3 +276,52 @@ func (s *tryUntilSuccSuite) TestShouldStopOnSuccess(c *C) {
...
+	c.Assert(err, ErrorMatches, "Fail")
+	c.Assert(callCount, Less, 10)
+
+	callCount = 0
+	err = RetryContext(ctx, 10, time.Millisecond, 2, func(ictx context.Context) error {
+		<-ictx.Done()
+		return errors.New("Canceled")
+	})
+	c.Assert(err, ErrorMatches, "Canceled")
+	c.Assert(callCount, Equals, 0)
```
Comment (kennytm) [inline pkg/util/util_test.go]: What's the purpose of this check? `callCount` is never modified inside lines 307-310.

Comment (suzaku) [reply pkg/util/util_test.go]: Fixed.

File: pump/node.go [hunk 2]
```diff
@@ -152,29 +152,54 @@ func (p *pumpNode) Notify(ctx context.Context) error {
...
 	}
 
 	return nil
 }
 
+func notifyDrainer(ctx context.Context, c *node.Status, dialerOpts []grpc.DialOption) error {
+	log.Info("Start trying to notify drainer", zap.String("addr", c.Addr))
+	var clientConn *grpc.ClientConn
+	err := util.RetryContext(ctx, 3, time.Second, 2, func(ictx context.Context) error {
+		log.Debug("Connecting drainer", zap.String("addr", c.Addr))
```
Comment (GregoryIan) [inline pump/node.go]:
```suggestion
		log.Info("Connecting drainer", zap.String("addr", c.Addr))
```

File: pump/node.go [hunk 3]
```diff
@@ -152,29 +152,54 @@ func (p *pumpNode) Notify(ctx context.Context) error {
...
+	})
+	if err != nil {
+		return errors.Annotatef(err, "connect drainer(%s)", c.Addr)
+	}
+	defer clientConn.Close()
+
+	drainer := pb.NewCisternClient(clientConn)
+
+	err = util.RetryContext(ctx, 3, time.Second, 2, func(ictx context.Context) error {
+		log.Debug("Notifying drainer", zap.String("addr", c.Addr))
```
Comment (GregoryIan) [inline pump/node.go]:
```suggestion
		log.Info("Notifying drainer", zap.String("addr", c.Addr))
```

**Comment by GregoryIan** [comment]
Rest LGTM

**Comment by GregoryIan** [comment]
@kennytm PTAL again

**Comment by GregoryIan** [comment]
LGTM

File: pump/node.go [hunk 4]
```diff
@@ -152,29 +152,54 @@ func (p *pumpNode) Notify(ctx context.Context) error {
...
 	return nil
 }
 
+func notifyDrainer(ctx context.Context, c *node.Status, dialerOpts []grpc.DialOption) error {
+	log.Info("Start trying to notify drainer", zap.String("addr", c.Addr))
+	var clientConn *grpc.ClientConn
+	err := util.RetryContext(ctx, 3, time.Second, 2, func(ictx context.Context) error {
+		log.Info("Connecting drainer", zap.String("addr", c.Addr))
+		var err error
+		clientConn, err = grpc.DialContext(ctx, c.Addr, dialerOpts...)
```
Comment (kennytm) [inline pump/node.go]: Should this use `ctx` or `ictx`?

File: pump/node.go [hunk 5]
```diff
@@ -152,29 +152,54 @@ func (p *pumpNode) Notify(ctx context.Context) error {
...
+	if err != nil {
+		return errors.Annotatef(err, "connect drainer(%s)", c.Addr)
+	}
+	defer clientConn.Close()
+
+	drainer := pb.NewCisternClient(clientConn)
+
+	err = util.RetryContext(ctx, 3, time.Second, 2, func(ictx context.Context) error {
+		log.Info("Notifying drainer", zap.String("addr", c.Addr))
+		_, err := drainer.Notify(ctx, nil)
```
Comment (kennytm) [inline pump/node.go]: ditto

**Pull request merged**